File: baekjoon/Gold/17298. 오큰수/오큰수.py
import sys
from collections import deque
input = sys.stdin.readline

n = int(input())
a = list(map(int, input().split()))
stack = []
answer = [-1]*n

for i in range(n):
    # 스택이 비어있지 않고, 스택의 가장 위에 있는 원소의 값이 현재 원소보다 작을 때까지 
    while stack and a[stack[-1]] < a[i]:
        # 스택의 가장 위에 있는 원소가 현재 원소보다 작다면, 해당 원소의 오큰수를 현재 원소로 갱신하고 스택에서 해당 원소를 제거
        answer[stack.pop()] = a[i]
    stack.append(i)
    
print(*answer)        

File: Baekjoon/Gold/17298. 오큰수/오큰수.java
import java.util.*;
import java.io.*;
public class Main {
  public static void main(String[] args) throws IOException {
    BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
    int n = Integer.parseInt(bf.readLine());
    int[]A = new int[n];    // 수열 배열 
    int[]ans = new int[n]; // 정답 배열 
    String[] str = bf.readLine().split(" ");
    for (int i = 0; i < n; i++) {
        A[i] = Integer.parseInt(str[i]);
    }
    Stack<Integer> myStack = new Stack<>();
    myStack.push(0); 
    for (int i = 1; i < n; i++) {
        
        while (!myStack.isEmpty() && A[myStack.peek()] < A[i]) {  
            ans[myStack.pop()] = A[i];  
        }
        myStack.push(i); 
    }
    while (!myStack.empty()) {
        ans[myStack.pop()] = -1;
    }
    BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    for (int i = 0; i < n; i++) {
        bw.write(ans[i] + " ");
        // 출력한다
    }
    bw.write("\n");
    bw.flush();
  }
}

